src/probability/ident_distrib.lean

```diff
@@ -0,0 +1,277 @@
+so on.
+-/
+
+open measure_theory filter finset
+
+noncomputable theory
+
+open_locale topological_space big_operators measure_theory ennreal nnreal
+
+variables {α β γ δ : Type*} [measurable_space α] [measurable_space β]
+  [measurable_space γ] [measurable_space δ]
+
+namespace probability_theory
+
+/-- Two functions defined on two (possibly different) measure spaces are identically distributed if
+their image measures coincide. This only makes sense when the functions are ae measurable
+(as otherwise the image measures are not defined), so we require this as well in the definition. -/
+structure ident_distrib
+  (f : α → γ) (g : β → γ) (μ : measure α . volume_tac)  (ν : measure β . volume_tac) : Prop :=
+(ae_measurable_fst : ae_measurable f μ)
+(ae_measurable_snd : ae_measurable g ν)
+(map_eq : measure.map f μ = measure.map g ν)
+
+namespace ident_distrib
+
+open topological_space
+
+variables {μ : measure α} {ν : measure β} {f : α → γ} {g : β → γ}
+
+protected lemma ae_measurable (h : ident_distrib f g μ ν) :
+  ae_measurable f μ := h.ae_measurable_fst
+
+protected lemma refl (hf : ae_measurable f μ) :
+  ident_distrib f f μ μ :=
+{ ae_measurable_fst := hf,
+  ae_measurable_snd := hf,
+  map_eq := rfl }
+
+protected lemma symm (h : ident_distrib f g μ ν) : ident_distrib g f ν μ :=
+{ ae_measurable_fst := h.ae_measurable_snd,
+  ae_measurable_snd := h.ae_measurable_fst,
+  map_eq := h.map_eq.symm }
+
+protected lemma trans {ρ : measure δ} {h : δ → γ}
+  (h₁ : ident_distrib f g μ ν) (h₂ : ident_distrib g h ν ρ) : ident_distrib f h μ ρ :=
+{ ae_measurable_fst := h₁.ae_measurable_fst,
+  ae_measurable_snd := h₂.ae_measurable_snd,
+  map_eq := h₁.map_eq.trans h₂.map_eq }
+
+protected lemma comp_of_ae_measurable {u : γ → δ} (h : ident_distrib f g μ ν)
+  (hu : ae_measurable u (measure.map f μ)) :
+  ident_distrib (u ∘ f) (u ∘ g) μ ν :=
+{ ae_measurable_fst := hu.comp_ae_measurable h.ae_measurable,
+  ae_measurable_snd :=
+    by { rw h.map_eq at hu, exact hu.comp_ae_measurable h.symm.ae_measurable },
+  map_eq :=
+  begin
+    rw [← ae_measurable.map_map_of_ae_measurable hu h.ae_measurable,
+      ← ae_measurable.map_map_of_ae_measurable _ h.symm.ae_measurable, h.map_eq],
+    rwa ← h.map_eq,
+  end }
+
+protected lemma comp {u : γ → δ} (h : ident_distrib f g μ ν) (hu : measurable u) :
+  ident_distrib (u ∘ f) (u ∘ g) μ ν :=
+h.comp_of_ae_measurable hu.ae_measurable
+
+lemma measure_mem_eq (h : ident_distrib f g μ ν) {s : set γ} (hs : measurable_set s) :
+  μ (f ⁻¹' s) = ν (g ⁻¹' s) :=
+by rw [← measure.map_apply_of_ae_measurable h.ae_measurable hs,
+  ← measure.map_apply_of_ae_measurable h.symm.ae_measurable hs, h.map_eq]
+
+lemma ae_mem_snd (h : ident_distrib f g μ ν) {t : set γ}
+  (tmeas : measurable_set t) (ht : ∀ᵐ x ∂μ, f x ∈ t) :
+   ∀ᵐ x ∂ν, g x ∈ t :=
+begin
+  apply (ae_map_iff h.symm.ae_measurable tmeas).1,
+  rw ← h.map_eq,
+  exact (ae_map_iff h.ae_measurable tmeas).2 ht,
+end
```

Comment:
Could you add another version of this with hypothesis and result of the form `∀ᵐ x ∂μ, p (f x)` for a predicate `p`? (and with hypothesis `measurable_set {x | p x}`)